app/code/community/Bolt/Boltpay/OrderCreationException.php

```diff
@@ -0,0 +1,214 @@
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_NOT_PURCHASABLE    = '{"reason": "The product is not purchasable", "product_id": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_GRAND_TOTAL        = '{"reason": "Grand total has changed", "old_value": "%d", "new_value": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_DISCOUNT           = '{"reason": "Discount total has changed", "old_value": "%d", "new_value": "%d"}';
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_TAX                = '{"reason": "Tax amount has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * Currently not used -- shipping tax in M1 is used to fix rounding problems and reports the
+     * full cart tax.  Therefore, checking full tax is sufficient
+     */
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_SHIPPING_TAX       = '{"reason": "Shipping tax amount has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * Item price changes
+     */
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED         = 2001004;
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED_TMPL    = '{"product_id": "%d", "old_price": "%d", "new_price": "%d"}';
+
+    /**
+     * Items have become out of stock without back-orders enabled
+     */
+    const E_BOLT_OUT_OF_INVENTORY         = 2001005;
+    const E_BOLT_OUT_OF_INVENTORY_TMPL    = '{"product_id": "%d", "available_quantity": %d, "needed_quantity": %d}';
+
+    /**
+     * Error applying discount to the cart
+     */
+    const E_BOLT_DISCOUNT_CANNOT_APPLY                 = 2001006;
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_EXPIRED    = '{"reason": "This coupon has expired", "discount_code": "%s"}';
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_GENERIC    = '{"reason": "%s", "discount_code": "%s"}';
+
+    /**
+     * Invalid discount code used
+     */
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST        = 2001007;
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST_TMPL   = '{"discount_code": "%s"}';
+
+    /**
+     * Change in shipping price
+     */
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED           = 2001008;
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED_TMPL      = '{"reason": "Shipping total has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * @var int http response code that is to be returned
+     */
+    protected $httpCode;
+
+    /**
+     * @var string The json to be returned to Bolt associated with this exception
+     */
+    protected $json;
+
+    /**
+     * @var array All possible error codes supported by Bolt
+     */
+    protected static $validCodes = array(
+        self::E_BOLT_GENERAL_ERROR,
+        self::E_BOLT_ORDER_ALREADY_EXISTS,
+        self::E_BOLT_CART_HAS_EXPIRED,
+        self::E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED,
+        self::E_BOLT_OUT_OF_INVENTORY,
+        self::E_BOLT_DISCOUNT_CANNOT_APPLY,
+        self::E_BOLT_DISCOUNT_DOES_NOT_EXIST,
+        self::E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED
+    );
+
+    /**
+     * Bolt_Boltpay_InvalidTransitionException constructor.
```

Comment:
> can you add unit test for `Bolt_Boltpay_OrderCreationException`. Only checking one specific error would be sufficient.

@daisy1754  
Yes.  I certainly will.

Proper unit test to check error the triggering of error codes will take an large amount of time.  It is not trivial and I do not want to approach it like is a trivial thing.  Checking one error code will not be sufficient and is at the heart of v2.0.0.  Even more, I prefer integration test and verifying the Bolt modal does what it is suppose to do.

Solid additional unit test will be done in a separate PR.
https://app.asana.com/0/941895179897714/1121207201774756

Reply:
Please, please don't treat unit testing as a separate step. It must be part of your change. If you add new meaningful logic, your PR must include unit test. I've kept saying this for long time. 

For other comments I'll give other team members a chance to jump in in case people have any opinion, but for this specific one my answer is clear - unit test has to be part of this PR.

Reply:
Oops, Kazuki, perhaps I was unclear.  I am certainly adding a unit test for one specific error for this PR that you feel is sufficient.  That is what I meant by, _"Yes, I certainly will."_

I am also saying that _"Solid additional unit test will be done in a separate PR."_ because this is at the heart of this project and its importance can not be overstated.  UI integration test also seem very important to me.

Your answer is clear.  Apologies for me being unclear.  Thank you for the solid and stern feedback.  It is certainly reflecting in an overall high quality of this release.

Reply:
@daisy1754 
I now understand where the confusion lies on my end.  You are looking at this from a micro perspective, (i.e. do the methods `getJson` and `getHttpCode` produce valid and expected results when  `Bolt_Boltpay_OrderCreationException` is constructed.)

My mind, at this point after so many beta releases and seeing the code perform at that level, is accustomed to be looking and thinking at the same thing from a macro level, (i.e. when manipulating the Magento system to produce the various error states, and being called from the validation methods, are all the correct JSON errors triggered and displayed by the the `Bolt_Boltpay_OrderCreationException` class.) 

With my understanding fixed, I'm sure now I can quickly produce what you are asking for.

Thanks again.  Your reviews have often been the ones to push me to deliver at an even higher-level and I am very grateful that you take the time for this.  Now, let's rise these PR's up to the next level and get this code into official production release!